arbitrum_types/txoptions.go

```diff
@@ -0,0 +1,110 @@
+package arbitrum_types
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/state"
+)
+
+type rejectedError struct {
+	msg string
+}
+
+func NewRejectedError(msg string) *rejectedError {
+	return &rejectedError{msg: msg}
+}
+func (e rejectedError) Error() string { return e.msg }
+func (rejectedError) ErrorCode() int  { return -32003 }
+
+type limitExceededError struct {
+	msg string
+}
+
+func NewLimitExceededError(msg string) *limitExceededError {
+	return &limitExceededError{msg: msg}
+}
+func (e limitExceededError) Error() string { return e.msg }
+func (limitExceededError) ErrorCode() int  { return -32005 }
+
+type RootHashOrSlots struct {
+	RootHash  *common.Hash
+	SlotValue map[common.Hash]common.Hash
+}
+
+func (r *RootHashOrSlots) UnmarshalJSON(data []byte) error {
+	var hash common.Hash
+	var err error
+	if err = json.Unmarshal(data, &hash); err == nil {
+		r.RootHash = &hash
+		return nil
+	}
+	return json.Unmarshal(data, &r.SlotValue)
+}
+
+func (r RootHashOrSlots) MarshalJSON() ([]byte, error) {
+	if r.RootHash != nil {
+		return json.Marshal(*r.RootHash)
+	}
+	return json.Marshal(r.SlotValue)
+}
+
+// TODO rename to just "Options" ?
+type ConditionalOptions struct {
+	KnownAccounts  map[common.Address]RootHashOrSlots `json:"knownAccounts"`
+	BlockNumberMin *hexutil.Uint64                    `json:"blockNumberMin,omitempty"`
+	BlockNumberMax *hexutil.Uint64                    `json:"blockNumberMax,omitempty"`
+	TimestampMin   *hexutil.Uint64                    `json:"timestampMin,omitempty"`
+	TimestampMax   *hexutil.Uint64                    `json:"timestampMax,omitempty"`
+}
+
+func (o *ConditionalOptions) Check(l1BlockNumber uint64, l1Timestamp uint64, statedb *state.StateDB) error {
+	if o.BlockNumberMin != nil && l1BlockNumber < uint64(*o.BlockNumberMin) {
+		return NewRejectedError("BlockNumberMin condition not met")
+	}
+	if o.BlockNumberMax != nil && l1BlockNumber > uint64(*o.BlockNumberMax) {
+		return NewRejectedError("BlockNumberMax condition not met")
+	}
+	if o.TimestampMin != nil && l1Timestamp < uint64(*o.TimestampMin) {
+		return NewRejectedError("TimestampMin condition not met")
+	}
+	if o.TimestampMax != nil && l1Timestamp > uint64(*o.TimestampMax) {
+		return NewRejectedError("TimestampMax condition not met")
+	}
+	return o.CheckOnlyStorage(statedb)
+}
+
+func (o *ConditionalOptions) PreCheck(l1BlockNumberLowerBound uint64, statedb *state.StateDB) error {
+	if o.BlockNumberMax != nil && l1BlockNumberLowerBound > uint64(*o.BlockNumberMax) {
+		return NewRejectedError("BlockNumberMax condition not met")
+	}
+	// TODO can we get a timestamp LowerBound?
+	//if o.TimestampMin != nil && l1TimestampLowerBound > uint64(*o.TimestampMax) {
```

Comment:
The last block's timestamp is a lower bound as timestamp can never decrease.

Reply:
this condition should also be `if o.TimestampMax != nil`

Reply:
do you mean L2 block timestamp? or L1 block?

L2 block timestamp is easily available in pre-checker which uses the `PreCheck` method, but I am not sure if L2 timestamp is good enough. My concern is if L1 block timestamp can be older then L2 block timestamp? If so, then L2 block timestamp isn't true lower bound of L1 block timestamp.

If you mean current L1 block timestamp, then I think that getting this value in pre-checker would require some additional operations. Maybe it could be tracked simillar to L1 block number

Reply:
L2 block timestamp, as that's what's visible in the EVM